openstack/identity/v3/applicationcredentials/requests.go

```diff
@@ -69,7 +72,25 @@ type CreateOpts struct {
 	// A list of access rules objects.
 	AccessRules []AccessRule `json:"access_rules,omitempty"`
 	// The expiration time of the application credential, if one was specified.
-	ExpiresAt string `json:"expires_at,omitempty"`
+	ExpiresAt *time.Time `json:"-"`
+}
+
+// MarshalJSON converts our struct request into JSON.
+func (opts CreateOpts) MarshalJSON() ([]byte, error) {
```

Comment:
This can instead be implemented as:

```go
// ToApplicationCredentialCreateMap formats a CreateOpts into a create request.
func (opts CreateOpts) ToApplicationCredentialCreateMap() (map[string]interface{}, error) {
  b, err := gophercloud.BuildRequestBody(opts, "application_credential")
  if err != nil {
    return b, err
  }

  if opts.ExpiresAt != nil {
    appCred := b["application_credential"].(map[string]interface{})
    appCred["expires_at"] = opts.ExpiresAt.Format(gophercloud.RFC3339Milli)
    b["application_credential"] = appCred
  }

  return b, nil
}
```

This pattern is used in other areas of Gophercloud already.

In addition, we'll need to verify what time format Keystone is expecting. We're using Milli here, but MilliNoZ is being returned - that might cause some confusion and usage errors.

Reply:
My idea was to avoid body modification inside the `To***CreateMap`, but do this in `MarshalJSON`.

I clarified time format:
* Trusts accept and return RFC3339Milli, but response truncates milliseconds. Therefore for tests we need to truncate milliseconds.
* AppCreds accept and return RFC3339MilliNoZ. However for tests we need to truncate nanoseconds.